sdk/core/core-util/test/public/reshape.spec.ts

```diff
@@ -0,0 +1,96 @@
+describe("reshape", function () {
+  it("renames a simple property", function () {
+    const input = { a: 1, b: 2 };
+    const actual = reshape(input, "a", "c");
+    const expected = { c: 1, b: 2 };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("maps a simple property", function () {
+    const input = { a: 1, b: 2 };
+    const actual = reshape(input, "a", String);
+    const expected = { a: "1", b: 2 };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("renames and maps a simple property", function () {
+    const input = { a: 1, b: 2 };
+    const actual = reshape(input, "a", "c", String);
+    const expected = { c: "1", b: 2 };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("renames a nested property", function () {
+    const input = { a: { c: { e: 4 }, d: 3 }, b: 2 };
+    const actual = reshape(input, "a.c.e", "f");
+    const expected = { a: { c: { f: 4 }, d: 3 }, b: 2 };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("can map elements of an array", function () {
+    const input = { a: { b: [1, 2, 3] } };
+    const actual = reshape(input, "a.b[]", String);
+    const expected = { a: { b: ["1", "2", "3"] } };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("renaming array elements is an error", function () {
+    const input = { a: { b: [1, 2, 3] } };
+    expect(() => reshape(input, "a.b[]", "oops")).toThrowError("cannot rename array indices");
+  });
+
+  it("can rename properties of elements of an array", function () {
+    const input = {
+      a: {
+        b: [
+          { c: 1, d: 2 },
+          { c: 3, d: 4 },
+        ],
+      },
+    };
+    const actual = reshape(input, "a.b[]c", "e");
+    const expected = {
+      a: {
+        b: [
+          { e: 1, d: 2 },
+          { e: 3, d: 4 },
+        ],
+      },
+    };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("selecting a property that doesn't exist does nothing", function () {
+    const input = { a: 1, b: 2 };
+    const actual = reshape(input, "c", "d");
+    const expected = { a: 1, b: 2 };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("selecting a property as an array that is not an array does nothing", function () {
+    const input = { a: 1, b: 2 };
+    const actual = reshape(input, "a[]", String);
+    const expected = { a: 1, b: 2 };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("property with a dot in the name can be selected", function () {
+    const input = { a: { ".c.": { e: 4 }, d: 3 }, b: 2 };
+    const actual = reshape(input, String.raw`a.\.c\..e`, "f");
+    const expected = { a: { ".c.": { f: 4 }, d: 3 }, b: 2 };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("property with a backslash in the name can be selected", function () {
+    const input = { a: { "\\c\\": { e: 4 }, d: 3 }, b: 2 };
+    const actual = reshape(input, String.raw`a.\\c\\.e`, "f");
+    const expected = { a: { "\\c\\": { f: 4 }, d: 3 }, b: 2 };
+    assert.deepEqual(actual, expected);
+  });
+});
```

Comment:
I wonder what would happened if we have an array of an arrray, { a: [[1, 2, 3]] }, if we use a[][] to select the elements, could it be interpret as [[1], [2], [3]] for lack of the length inforamation of the first dimension ?

Reply:
It would recurse through the nested arrays correctly, but this exposed a small issue where I didn't handle chaining array selectors correctly. I updated the code and added a test case for this. Good question!